### PR TITLE
FEATURE: filter personal messages by tags

### PR DESCRIPTION
**Pull request by techAPJ**
This PR adds a tag dropdown on user messages page to filter messages by tags.

screen shot 2018-03-07 at 16 40 37

### CODE REVIEW

**Comment by techAPJ** [comment]
Build failure seems irrelevant.

**Comment by jjaffeux** [comment]
@techAPJ I wonder if that wouldn‘t be better with a specific {{pm-tag-drop}} component:

```
// select-kit/components/pm-tag-drop.js.es6
import TagDropComponent from "select-kit/components/tag-drop";
import { computed } from "ember-addons/ember-computed-decorators";

export default TagDropComponent.extend({
  @computed
  allTagsUrl() {
    return `/u/${this.currentUser.username}/messages/`;
  },

  content: Ember.computed.alias("pmTags"),

  actions: {
    onSelect(tagId) {
      const url = `/u/${this.currentUser.username}/messages/tag/${tagId}`;
      DiscourseURL.routeTo(url);
    }
  }
```

Usage:

```
{{pm-tag-drop pmTags=pmTags}}
```

Didn‘t try it on my machine, let me know if you have issues with this.

**Comment by techAPJ** [comment]
Thanks for the review @jjaffeux and @tgxworld.

Made the changes/additions as per feedback.

File: app/models/tag.rb [hunk 1]
```diff
@@ -59,6 +59,24 @@ def self.top_tags(limit_arg: nil, category: nil, guardian: nil)
...
+      FROM tags
+      INNER JOIN topic_tags ON tags.id = topic_tags.tag_id
+      INNER JOIN topics ON topics.id = topic_tags.topic_id
+      AND topics.deleted_at IS NULL
+      AND topics.archetype = 'private_message'
+      GROUP BY tags.name
+      LIMIT #{limit}
+    SQL
+
+    tag_names.map { |row| row['tag_name'] }
```
Comment (tgxworld) [inline app/models/tag.rb]: I think `tag_names`.values would work

File: app/models/topic_list.rb [hunk 2]
```diff
@@ -59,6 +60,15 @@ def top_tags
     Tag.top_tags(opts)
   end
 
+  def pm_tags
+    guardian = Guardian.new(@current_user)
+    if @opts[:show_pm_tags] && guardian.can_tag_pms?
+      Tag.pm_tags(guardian: guardian)
+    else
+      return []
```
Comment (tgxworld) [inline app/models/topic_list.rb]: no need for this `return` :)

File: spec/models/topic_list_spec.rb [hunk 3]
```diff
@@ -78,4 +78,28 @@
...
+
+    context 'when viewed as normal user' do
+      it 'returns no tags' do
+        expect(TopicList.new('liked', personal_message.user, [personal_message], {show_pm_tags: true}).pm_tags).to be_empty
+      end
+    end
+
+    context 'when viewed as admin' do
+      it 'returns pm tags' do
+        expect(TopicList.new('liked', admin, [personal_message], {show_pm_tags: true}).pm_tags).to contain_exactly(@private_tags.map(&:name)[0], @private_tags.map(&:name)[1])
```
Comment (tgxworld) [inline spec/models/topic_list_spec.rb]: Oops my bad, in this case I think we can just use `match_array` since the order doesn't matter.

**Comment by tgxworld** [comment]
Nice work @techAPJ  :+1:

**Comment by techAPJ** [comment]
@tgxworld Thanks! Changes made. Will merge once the Travis build is green.

**Comment by jjaffeux** [comment]
awesome work

**Pull request merged**